### PR TITLE
URL Rewriting

### LINKED ISSUE #200: [url rewriting.xml] URL Rewrite article needs refactoring
The URL rewriting article needs refactoring. 

URL rewriting is a very important mechanism in eXist for application building and I think it deserves better. 

Take readers in more slowly, postpone stuff about servlets and MVC patterns and all stuff you don't need for the basic understanding to sections at the end, update text *and examples* to the current conventions in running code from the db and not from the fs, etc.

This is a major job. I know because I wrote the chapter about this in the eXist book that took me several days. 

*   exist-db version: `4.0.0`
*   documentation version: `4.0.0`

### PR DESCRIPTION
**Pull request by marmoure**
Closes #200 .
This PR uses @wolfgangmm #227  for the URL Rewriting and all the fixes in that Pr.
It removes outdated parts, rename all `xql` to `xq`.

This open source contribution to the [documentation](https://github.com/eXist-db/documentation) project was commissioned by the Office of the Historian, U.S. Department of State, https://history.state.gov/.

### CODE REVIEW

**Comment by joewiz** [comment]
Thank you, @marmoure and @adamretter. 

With the stated goal of #200 being a thoroughly updated article that is approachable and presents the key information that users need to know about the URL Rewriting facility (a "major job", as Erik characterized it), I'd suggest that more revision is necessary here. Wolfgang's draft revised introduction is a good start, but the body of the article needs to be reviewed to meet the stated goals of #200. As a starting point, I would suggest reading the relevant sections from the eXist book (pp. 194-207), since it comprehensively describes the URL rewriting facility. Make a note of what's covered there - for example, permissions considerations - and be sure all information is addressed here and introduced coherently. As Erik recommended, "postpone stuff about servlets and MVC patterns and all stuff you don't need for the basic understanding to sections at the end." The article should address the typical user looking to exert full control over their URLs and tell them what they need to know to do so.

Also, please be sure the CI passes.

**Comment by yamahito** [comment]
@joewiz I've spent some time reworking the article, restructuring in some places, adding examples, and aiming for a more concise wording.  How does it look?

File: src/main/xar-resources/data/urlrewrite/urlrewrite.xml [hunk 1]
```diff
@@ -372,7 +422,7 @@
       throughout the entire pipeline, we need to clear the
       <literal>xslt.input</literal>
       with an explicit call to clear-attribute.</para>
-    <para>What benefits does it have to exchange data through request attributes: We save one serialization step:
+    <para>The benefit does to exchanging data through request attributes is that we save one serialization step.
```
Comment (line-o) [inline src/main/xar-resources/data/urlrewrite/urlrewrite.xml:425]: I suggest:

`The benefit of exchanging data through request attributes is that we save one serialization step.`

**Comment by line-o** [comment]
@yamahito much clearer now 👍🏼

File: src/main/xar-resources/data/urlrewrite/urlrewrite.xml [hunk 2]
```diff
@@ -1,6 +1,4 @@
-<?xml-model href="http://docbook.org/xml/5.0/rng/docbook.rng" schematypens="http://relaxng.org/ns/structure/1.0"?>
-<?xml-model href="http://docbook.org/xml/5.0/rng/docbook.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
-<article version="5.0" xmlns="http://docbook.org/ns/docbook" xmlns:xlink="http://www.w3.org/1999/xlink">
+<?xml-model href="http://docbook.org/xml/5.0/rng/docbook.rng" schematypens="http://relaxng.org/ns/structure/1.0"?><?xml-model href="http://docbook.org/xml/5.0/rng/docbook.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?><article xmlns="http://docbook.org/ns/docbook" xmlns:xlink="http://www.w3.org/1999/xlink" version="5.0">
   <info>
     <title>URL Rewriting</title>
     <date>2Q19</date>
```
Comment (joewiz) [inline src/main/xar-resources/data/urlrewrite/urlrewrite.xml:4]: According to the [Author Reference](https://www.exist-db.org/exist/apps/doc/author-reference#info), this should be updated to February 2023 or 1Q23.

File: src/main/xar-resources/data/urlrewrite/urlrewrite.xml [hunk 3]
```diff
@@ -10,175 +8,169 @@
...
         </listitem>
-      </varlistentry>
-      <varlistentry>
-        <term>
-          <code>exist:root</code>
-        </term>
+      </orderedlist>
+    <sect2 xml:id="eg1">
+      <title>Example I: A Simple Implementation</title>
+      <para>Consider  the document you are currently reading; a direct URL pointing to the source data might be <code>/exist/apps/doc/data/urlrewrite.xml</code>.  But accessing this URI  would only show a user the raw XML: instead users should get a properly formatted HTML version of the text, accessible through a simple URL like <code>/exist/apps/doc/urlrewrite</code>.</para>
```
Comment (joewiz) [inline src/main/xar-resources/data/urlrewrite/urlrewrite.xml:34]: I think the direct source URL to the source XML for this article would be `/exist/rest/db/apps/doc/data/urlrewrite.xml`.

File: src/main/xar-resources/data/urlrewrite/urlrewrite.xml [hunk 4]
```diff
@@ -10,175 +8,169 @@
...
+          <para><code>XQueryURLRewrite</code> uses the controller XML fragment as an instruction on what to do next.  This instruction may be as simple as forwarding to a resource on (or off) the server, or it may be as complex as a pipeline using the <emphasis>Model-View-Controller</emphasis> pattern (see <xref linkend="mvc-pipelines"/>) and other servlets such as eXist's <xref linkend="xq-servlet"/> or <xref linkend="xslt-servlet"/>.</para>
         </listitem>
-      </varlistentry>
-      <varlistentry>
-        <term>
-          <code>exist:root</code>
-        </term>
+      </orderedlist>
+    <sect2 xml:id="eg1">
+      <title>Example I: A Simple Implementation</title>
```
Comment (joewiz) [inline src/main/xar-resources/data/urlrewrite/urlrewrite.xml:33]: The next example is "Example 2", so perhaps best to stick with Arabic numerals -> "Example 1".

File: src/main/xar-resources/data/urlrewrite/listings/listing-1.txt [hunk 5]
```diff
@@ -0,0 +1,13 @@
+xquery version "3.1";
+
+declare variable $exist:path external;
+declare variable $exist:resource external;
+declare variable $exist:controller external;
+declare variable $exist:prefix external;
+declare variable $exist:root external;
+
+<dispatch xmlns="http://exist.sourceforge.net/NS/exist">
+    <forward url="{$exist:controller}/modules/transform.xql">
```
Comment (joewiz) [inline src/main/xar-resources/data/urlrewrite/listings/listing-1.txt:10]: The text refers to this XQuery as `transform.xq` - i.e., without the `l`.

File: src/main/xar-resources/data/urlrewrite/listings/listing-2.txt [hunk 6]
```diff
@@ -1,8 +1,14 @@
...
+declare variable $exist:path external;
+declare variable $exist:resource external;
+declare variable $exist:controller external;
+declare variable $exist:prefix external;
+declare variable $exist:root external;
+
+<dispatch xmlns="http://exist.sourceforge.net/NS/exist">
+    <forward url="data/{$exist:resource}.xml"/>
+    <view>
+        <forward url="{$exist:controller}/modules/transform.xql"/>
```
Comment (joewiz) [inline src/main/xar-resources/data/urlrewrite/listings/listing-2.txt:12]: The text refers to this XQuery as `transform.xq` - i.e., without the `l`.

File: src/main/xar-resources/data/urlrewrite/urlrewrite.xml [hunk 7]
```diff
@@ -10,175 +8,169 @@
...
+      <para>For example, one may want to direct all requests to <literal>/exist/apps/doc/{resource}</literal> to an XQuery, <literal>transform.xq</literal>, which is responsible for converting the XML content into HTML:</para>
+      <programlisting language="xquery" xlink:href="listings/listing-1.txt"/>
+      <para>This example controller returns a simple <tag>dispatch</tag> fragment which will be passed back to the URL rewriting framework. The <tag>forward</tag> element instructs the framework to call the URL <literal>modules/transform.xq</literal> relative to the collection in which the controller resides. It adds a request parameter, named <literal>doc</literal>, which translates the requested resource into an actual document path to be transformed.</para>
```
Comment (joewiz) [inline src/main/xar-resources/data/urlrewrite/urlrewrite.xml:39]: Consider adding another sentence:
> The receiving query could access this parameter using `request:get-parameter("doc", ())` in order to retrieve the requested article.

Comment (joewiz) [inline src/main/xar-resources/data/urlrewrite/urlrewrite.xml:39]: I'm not sure it's totally accurate to say that the request parameter named `doc` "translates the requested resource into an actual document path". Instead, `doc` contains the name of the resource sans path components (e.g., `urlrewrite`) plus the file extension of the actual document stored in the database (e.g., `urlrewrite.xml`).

File: src/main/xar-resources/data/urlrewrite/urlrewrite.xml (hunk 2 above)
Comment (yamahito) [reply src/main/xar-resources/data/urlrewrite/urlrewrite.xml:4]: Yes; will fix that.

File: src/main/xar-resources/data/urlrewrite/urlrewrite.xml (hunk 3 above)
Comment (yamahito) [reply src/main/xar-resources/data/urlrewrite/urlrewrite.xml:34]: I will reword this, since the URLs do not really point to this exact document (which is at `/exist/apps/doc/data/urlrewrite/urlrewrite.xml` - I think bringing the rest interface into it muddies the waters even further).  I considered changing all of the URLs to match the actual docbook article, but I think the important thing is that they are consistent.

**Comment by yamahito** [comment]
@joewiz I've attempted to address your feedback points.  Thanks!

File: src/main/xar-resources/data/urlrewrite/urlrewrite.xml [hunk 8]
```diff
@@ -1,193 +1,300 @@
-<?xml-model href="http://docbook.org/xml/5.0/rng/docbook.rng" schematypens="http://relaxng.org/ns/structure/1.0"?>
-<?xml-model href="http://docbook.org/xml/5.0/rng/docbook.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
-<article version="5.0" xmlns="http://docbook.org/ns/docbook" xmlns:xlink="http://www.w3.org/1999/xlink">
+<?xml-model href="http://docbook.org/xml/5.0/rng/docbook.rng" schematypens="http://relaxng.org/ns/structure/1.0"?><?xml-model href="http://docbook.org/xml/5.0/rng/docbook.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?><article xmlns="http://docbook.org/ns/docbook" xmlns:xlink="http://www.w3.org/1999/xlink" version="5.0">
   <info>
     <title>URL Rewriting</title>
-    <date>2Q19</date>
+    <date>2022-02</date>
```
Comment (joewiz) [inline src/main/xar-resources/data/urlrewrite/urlrewrite.xml:4]: It's 2023 ;)

Comment (joewiz) [reply src/main/xar-resources/data/urlrewrite/urlrewrite.xml:4]: Fixed in a PR against this branch - https://github.com/evolvedbinary/documentation/pull/1.

Comment (adamretter) [reply src/main/xar-resources/data/urlrewrite/urlrewrite.xml]: Should be fixed now.

**Comment by adamretter** [comment]
@joewiz I have fixed the date and cleaned up the commit history. This should be ready for you to merge now if you are happy to do so.

**Comment by joewiz** [comment]
@adamretter Would you please pull in my https://github.com/evolvedbinary/documentation/pull/1, so that this PR incorporates that review?

**Comment by adamretter** [comment]
@joewiz Sorry, I wasn't aware of your https://github.com/evolvedbinary/documentation/pull/1.
I think it''s best not to send PRs to our fork please.

I just took a look at that PR and there seems to be merge conflicts. Can I suggest that this gets merged first, and then you could reopen your PR against this repo.

**Comment by joewiz** [comment]
@adamretter Ok, that'll work.

**Review by joewiz — APPROVED** [review]
@yamahito Thanks, this is a great article!
